django/db/migrations/graph.py

```diff
@@ -5,6 +5,62 @@
 from django.db.migrations.state import ProjectState
 from django.utils.datastructures import OrderedSet
 from django.utils.encoding import python_2_unicode_compatible
+from django.utils.functional import cached_property, total_ordering
+
+
+@python_2_unicode_compatible
+@total_ordering
+class Node(object):
+    """
+    A single node in the migration graph. Contains direct links to adjacent
+    nodes in either direction.
+    """
+    def __init__(self, key):
+        self.key = key
+        self.children = set()
+        self.parents = set()
+
+    def __eq__(self, other):
+        return self.key == other
+
+    def __lt__(self, other):
+        return self.key < other
+
+    def __hash__(self):
+        return hash(self.key)
+
+    def __getitem__(self, item):
+        return self.key[item]
+
+    def add_parent(self, parent):
+        self.parents.add(parent)
+
+    def add_child(self, child):
+        self.children.add(child)
+
+    @cached_property
+    def ancestors(self):
+        # Use self.key instead of self to speed up the frequent hashing
+        # when constructing an OrderedSet.
+        ancestors = deque([self.key])
+        for parent in sorted(self.parents):
+            ancestors.extendleft(reversed(parent.ancestors))
+        return list(OrderedSet(ancestors))
+
+    @cached_property
+    def descendants(self):
+        # Use self.key instead of self to speed up the frequent hashing
+        # when constructing an OrderedSet.
+        descendants = deque([self.key])
+        for child in sorted(self.children):
+            descendants.extendleft(reversed(child.descendants))
+        return list(OrderedSet(descendants))
+
+    def __str__(self):
```

Comment:
Style: can you sort `__str__` and `__repr__` between the other private methods on top.